File: MiniDOMOffice/Office/Mail/Application/MailRootFolder.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using minidom;
using minidom.repositories;
using static minidom.Sistema;
using static minidom.Anagrafica;
using static minidom.Office;


namespace minidom
{
    public partial class Office
    {

        /// <summary>
        /// Root folder di un oggetto <see cref="MailApplication"/>
        /// </summary>
        [Serializable]
        public class MailRootFolder
            : MailFolder
        {
            [NonSerialized] private MailFolder m_Inbox;
            [NonSerialized] private MailFolder m_TrashBin;
            [NonSerialized] private MailFolder m_Drafts;
            [NonSerialized] private MailFolder m_Sent;
            [NonSerialized] private MailFolder m_Spam;
            [NonSerialized] private MailFolder m_Archive;
            [NonSerialized] private MailFolder m_FindFolder;

            /// <summary>
            /// Costruttore
            /// </summary>
            public MailRootFolder()
            {
                m_Inbox = null;
                m_TrashBin = null;
                m_Drafts = null;
                m_Sent = null;
                m_Spam = null;
                m_Archive = null;
                m_FindFolder = null;
            }

            /// <summary>
            /// Cartella archivio
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public MailFolder Archive
            {
                get
                {
                    if (m_Archive is null)
                        m_Archive = Childs.GetItemByName("/archive");
                    if (m_Archive is null)
                    {
                        m_Archive = Childs.Add("archive");
                        m_Archive.Stato = ObjectStatus.OBJECT_VALID;
                        m_Archive.Save();
                    }

                    return m_Archive;
                }
            }

            /// <summary>
            /// Cartelle ricerche
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public MailFolder FindFolder
            {
                get
                {
                    if (m_FindFolder is null)
                        m_FindFolder = Childs.GetItemByName("/findfolder");
                    if (m_FindFolder is null)
                    {
                        m_FindFolder = Childs.Add("findfolder");
                        m_FindFolder.Stato = ObjectStatus.OBJECT_VALID;
                        m_FindFolder.Save();
                    }

                    return m_FindFolder;
                }
            }

            /// <summary>
            /// Cartella principale della posta in arrivo
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public MailFolder Inbox
            {
                get
                {
                    if (m_Inbox is null)
                        m_Inbox = Childs.GetItemByName("/inbox");
                    if (m_Inbox is null)
                    {
                        m_Inbox = Childs.Add("inbox");
                        m_Inbox.Stato = ObjectStatus.OBJECT_VALID;
                        m_Inbox.Save();
                    }

                    return m_Inbox;
                }
            }

            /// <summary>
            /// Cartella principale del cestino
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public MailFolder TrashBin
            {
                get
                {
                    if (m_TrashBin is null)
                        m_TrashBin = Childs.GetItemByName("/recycler");
                    if (m_TrashBin is null)
                    {
                        m_TrashBin = Childs.Add("recycler");
                        m_TrashBin.Stato = ObjectStatus.OBJECT_VALID;
                        m_TrashBin.Save();
                    }

                    return m_TrashBin;
                }
            }

            /// <summary>
            /// Cartella delle bozze
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public MailFolder Drafts
            {
                get
                {
                    if (m_Drafts is null)
                        m_Drafts = Childs.GetItemByName("/drafts");
                    if (m_Drafts is null)
                    {
                        m_Drafts = Childs.Add("drafts");
                        m_Drafts.Stato = ObjectStatus.OBJECT_VALID;
                        m_Drafts.Save();
                    }

                    return m_Drafts;
                }
            }

            /// <summary>
            /// Cartella delle email inviate
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public MailFolder Sent
            {
                get
                {
                    if (m_Sent is null)
                        m_Sent = Childs.GetItemByName("/sent");
                    if (m_Sent is null)
                    {
                        m_Sent = Childs.Add("sent");
                        m_Sent.Stato = ObjectStatus.OBJECT_VALID;
                        m_Sent.Save();
                    }

                    return m_Sent;
                }
            }

            /// <summary>
            /// Cartella delle email indesiderate
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public MailFolder Spam
            {
                get
                {
                    if (m_Spam is null)
                        m_Spam = Childs.GetItemByName("/spam");
                    if (m_Spam is null)
                    {
                        m_Spam = Childs.Add("spam");
                        m_Spam.Stato = ObjectStatus.OBJECT_VALID;
                        m_Spam.Save();
                    }

                    return m_Spam;
                }
            }

            /// <summary>
            /// Aggiorna la cartella
            /// </summary>
            /// <param name="folder"></param>
            /// <returns></returns>
            protected internal override MailFolder UpdateFolder(MailFolder folder)
            {
                if (folder is null)
                    throw new ArgumentNullException("folder");
                if (DBUtils.GetID(folder, 0) == 0)
                    return null;

                if (DBUtils.GetID(folder, 0) == DBUtils.GetID(Inbox, 0))
                {
                    m_Inbox = folder;
                }
                else if (DBUtils.GetID(folder, 0) == DBUtils.GetID(m_Archive, 0))
                {
                    m_Archive = folder;
                }
                else if (DBUtils.GetID(folder, 0) == DBUtils.GetID(m_TrashBin, 0))
                {
                    m_TrashBin = folder;
                }
                else if (DBUtils.GetID(folder, 0) == DBUtils.GetID(m_Drafts, 0))
                {
                    m_Drafts = folder;
                }
                else if (DBUtils.GetID(folder, 0) == DBUtils.GetID(m_Sent, 0))
                {
                    m_Sent = folder;
                }
                else if (DBUtils.GetID(folder, 0) == DBUtils.GetID(m_Spam, 0))
                {
                    m_Spam = folder;
                }
                else if (DBUtils.GetID(folder, 0) == DBUtils.GetID(m_FindFolder, 0))
                {
                    m_FindFolder = folder;
                }

                return base.UpdateFolder(folder);
            }
        }
    }
}

File: MiniDOMOffice/Office/Mail/Application/MailRootFolder.vb
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports System.Text.RegularExpressions
Imports System.Net.Mail
Imports System.Net.Mime
Imports minidom.Net.Mime
Imports minidom.Sistema
Imports minidom.Databases
Imports minidom

Partial Class Office

    <Serializable>
    Public Class MailRootFolder
        Inherits MailFolder

        Private m_Inbox As MailFolder
        Private m_TrashBin As MailFolder
        Private m_Drafts As MailFolder
        Private m_Sent As MailFolder
        Private m_Spam As MailFolder
        Private m_Archive As MailFolder
        Private m_FindFolder As MailFolder

        Public Sub New()
            Me.m_Inbox = Nothing
            Me.m_TrashBin = Nothing
            Me.m_Drafts = Nothing
            Me.m_Sent = Nothing
            Me.m_Spam = Nothing
            Me.m_Archive = Nothing
            Me.m_FindFolder = Nothing
        End Sub

        ''' <summary>
        ''' Cartella archivio
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property Archive As MailFolder
            Get
                If (Me.m_Archive Is Nothing) Then Me.m_Archive = Me.Childs.GetItemByName("/archive")
                If (Me.m_Archive Is Nothing) Then
                    Me.m_Archive = Me.Childs.Add("archive")
                    Me.m_Archive.Stato = ObjectStatus.OBJECT_VALID
                    Me.m_Archive.Save()
                End If
                Return Me.m_Archive
            End Get
        End Property

        ''' <summary>
        ''' Cartelle ricerche
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property FindFolder As MailFolder
            Get
                If (Me.m_FindFolder Is Nothing) Then Me.m_FindFolder = Me.Childs.GetItemByName("/findfolder")
                If (Me.m_FindFolder Is Nothing) Then
                    Me.m_FindFolder = Me.Childs.Add("findfolder")
                    Me.m_FindFolder.Stato = ObjectStatus.OBJECT_VALID
                    Me.m_FindFolder.Save()
                End If
                Return Me.m_FindFolder
            End Get
        End Property

        ''' <summary>
        ''' Cartella principale della posta in arrivo
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property Inbox As MailFolder
            Get
                If (Me.m_Inbox Is Nothing) Then Me.m_Inbox = Me.Childs.GetItemByName("/inbox")
                If (Me.m_Inbox Is Nothing) Then
                    Me.m_Inbox = Me.Childs.Add("inbox")
                    Me.m_Inbox.Stato = ObjectStatus.OBJECT_VALID
                    Me.m_Inbox.Save()
                End If
                Return Me.m_Inbox
            End Get
        End Property

        ''' <summary>
        ''' Cartella principale del cestino
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property TrashBin As MailFolder
            Get
                If (Me.m_TrashBin Is Nothing) Then Me.m_TrashBin = Me.Childs.GetItemByName("/recycler")
                If (Me.m_TrashBin Is Nothing) Then
                    Me.m_TrashBin = Me.Childs.Add("recycler")
                    Me.m_TrashBin.Stato = ObjectStatus.OBJECT_VALID
                    Me.m_TrashBin.Save()
                End If
                Return Me.m_TrashBin
            End Get
        End Property

        ''' <summary>
        ''' Cartella delle bozze
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property Drafts As MailFolder
            Get
                If (Me.m_Drafts Is Nothing) Then Me.m_Drafts = Me.Childs.GetItemByName("/drafts")
                If (Me.m_Drafts Is Nothing) Then
                    Me.m_Drafts = Me.Childs.Add("drafts")
                    Me.m_Drafts.Stato = ObjectStatus.OBJECT_VALID
                    Me.m_Drafts.Save()
                End If
                Return Me.m_Drafts
            End Get
        End Property

        ''' <summary>
        ''' Cartella delle email inviate
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property Sent As MailFolder
            Get
                If (Me.m_Sent Is Nothing) Then Me.m_Sent = Me.Childs.GetItemByName("/sent")
                If (Me.m_Sent Is Nothing) Then
                    Me.m_Sent = Me.Childs.Add("sent")
                    Me.m_Sent.Stato = ObjectStatus.OBJECT_VALID
                    Me.m_Sent.Save()
                End If
                Return Me.m_Sent
            End Get
        End Property

        ''' <summary>
        ''' Cartella delle email indesiderate
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property Spam As MailFolder
            Get
                If (Me.m_Spam Is Nothing) Then Me.m_Spam = Me.Childs.GetItemByName("/spam")
                If (Me.m_Spam Is Nothing) Then
                    Me.m_Spam = Me.Childs.Add("spam")
                    Me.m_Spam.Stato = ObjectStatus.OBJECT_VALID
                    Me.m_Spam.Save()
                End If
                Return Me.m_Spam
            End Get
        End Property

        Protected Friend Overrides Function UpdateFolder(folder As MailFolder) As MailFolder
            If (folder Is Nothing) Then Throw New ArgumentNullException("folder")
            If (GetID(folder) = 0) Then Return Nothing
            If (GetID(folder) = GetID(Me.Inbox)) Then
                Me.m_Inbox = folder
            ElseIf (GetID(folder) = GetID(Me.m_Archive)) Then
                Me.m_Archive = folder
            ElseIf (GetID(folder) = GetID(Me.m_TrashBin)) Then
                Me.m_TrashBin = folder
            ElseIf (GetID(folder) = GetID(Me.m_Drafts)) Then
                Me.m_Drafts = folder
            ElseIf (GetID(folder) = GetID(Me.m_Sent)) Then
                Me.m_Sent = folder
            ElseIf (GetID(folder) = GetID(Me.m_Spam)) Then
                Me.m_Spam = folder
            ElseIf (GetID(folder) = GetID(Me.m_FindFolder)) Then
                Me.m_FindFolder = folder
            End If
            Return MyBase.UpdateFolder(folder)
        End Function


    End Class

End Class